### PR TITLE
lambda-promtail: Add support for VPC flow Logs to lambda-promtail

### LINKED ISSUE #6767: Ability for Lambda Promtail Process VPC flow logs in stored S3 (or advice on how I may write it)
**Is your feature request related to a problem? Please describe.**
Not a problem

**Describe the solution you'd like**
The ability for Lambda promtail to process VPC flow logs in S3 buckets.

**Describe alternatives you've considered**
Using Cloudwatch instead. The motivation for not using CW is to save money, there is no reason to pay for ingest fees if we are planning to access the data via loki.

**Considering how to write it**
From looking at `s3.go` would it be as simple as adding a parameter to `processS3Event()` where there can be an option to process load balancer or VPC flow logs? Additional regex can then be added `var()`

Or is it best to essentially have two sets of label gathering logic in lines 106-111: https://github.com/grafana/loki/blob/d38c208aa73e25311b9ec83fce02553dc70cc4b8/tools/lambda-promtail/lambda-promtail/s3.go#L106-L111   and then concatenate the load-balancer and VPC flow log labels gathered into `labels` and return it?

### PR DESCRIPTION
**Pull request by chriskuchin**
This add support to the S3 parsing logic for AWS VPC Flow logs.

It also makes a small change to allow not printing log lines during processing. It leaves the default of printing the log lines in place.

**Which issue(s) this PR fixes**:
Fixes #6767

**Special notes for your reviewer**:

**Checklist**
- [x] Reviewed the `CONTRIBUTING.md` guide
- [x] Documentation added
- [x] Tests updated
- [x] `CHANGELOG.md` updated
- [x] Changes that require user attention or interaction to upgrade are documented in `docs/sources/upgrading/_index.md`

### CODE REVIEW

File: tools/lambda-promtail/lambda-promtail/main.go [hunk 1]
```diff
@@ -83,6 +84,12 @@ func setupArguments() {
 		batchSize, _ = strconv.Atoi(batch)
 	}
 
+	print := os.Getenv("PRINT_LOG_LINE")
```
Comment (kavirajk) [inline tools/lambda-promtail/lambda-promtail/main.go:87]: Any specific rationale behind naming of this ENV (`PRINT_LOG_LINE`)?. 

I would prefer to name it more descriptive `PROMTAIL_LAMBDA_PRINT_LOG_LINE`.

File: tools/lambda-promtail/lambda-promtail/s3_test.go [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+			if !reflect.DeepEqual(got, tt.want) {
+				t.Errorf("getLabels() = %v, want %v", got, tt.want)
+			}
+		})
+	}
+}
+
+func Test_parseS3Log(t *testing.T) {
+	type args struct {
+		ctx       context.Context
```
Comment (kavirajk) [inline tools/lambda-promtail/lambda-promtail/s3_test.go:107]: nit: I'm not seeing it's used differently for different tests. Can we remove it from `args` and have one `ctx`?

File: tools/lambda-promtail/lambda-promtail/s3_test.go [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+					"type": LB_LOG_TYPE,
+				},
+			},
+			wantErr: false,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			batchSize = tt.args.batchSize
+			tt.args.obj, _ = os.Open(tt.args.filename)
```
Comment (kavirajk) [inline tools/lambda-promtail/lambda-promtail/s3_test.go:153]: Can we not miss the error check here please?. Can lead to weird panics and brittle test on the long run IMO.

File: tools/lambda-promtail/lambda-promtail/main.go (hunk 1 above)
Comment (kavirajk) [reply tools/lambda-promtail/lambda-promtail/main.go:87]: Ah. never mind. I see other envs also doesn't take any prefixes :(

**Review by kavirajk — COMMENTED** [review]
thanks @chriskuchin. Overall LGTM. Added few minor nits.

One thing need to be done before merging this PR is to update the lambda promtail docs
https://grafana.com/docs/loki/latest/clients/lambda-promtail/

That should mention the support for vpc flow logs and should explain how to use them. Can you please take care of that as well?

File: tools/lambda-promtail/lambda-promtail/s3_test.go (hunk 3 above)
Comment (chriskuchin) [reply tools/lambda-promtail/lambda-promtail/s3_test.go]: Added an Error check and message to point out the problem

File: tools/lambda-promtail/lambda-promtail/s3_test.go (hunk 2 above)
Comment (chriskuchin) [reply tools/lambda-promtail/lambda-promtail/s3_test.go]: Is just passing in the `context.Background()` ok?

**Comment by chriskuchin** [comment]
Added a small note in the docs folder about the flow logs capability. Also made note that the timestamp on the logs won't be the actual log timestamp since flow logs don't have a timestamp by default

**Comment by kavirajk** [comment]
@chriskuchin can you rebase it with `main`? Suddenly it shows `555 changed files`

**Comment by chriskuchin** [comment]
hmm I am struggling on this rebase. getting ready to take off will look again later

**Comment by chriskuchin** [comment]
@kavirajk Moved to my main branch and opened a new MR that doesn't see to have the problems of the 600 fiels

https://github.com/grafana/loki/pull/7891

**Comment by chriskuchin** [comment]
@kavirajk Figured it out 😮‍💨 Needed to force push the rebase

**Comment by chriskuchin** [comment]
@kavirajk ok the rebase looks correct now but there is a failing test that doesn't seem related to my changes. Let me know if I'm wrong not great at reading go test output

**Comment by chriskuchin** [comment]
@kavirajk Looks like tests are finally passing!!!

**Comment by cstyan** [comment]
I'll review soon! Curious if https://github.com/grafana/loki/pull/8066 would help you as well

**Comment by chriskuchin** [comment]
@cstyan Really appreciate the review. I could see generic log parsing being useful, maybe. However, at this point the only logs I am targeting are ALB and Flow logs so it's not a high priority for me as we don't have any other logs in s3 currently.

**Comment by chriskuchin** [comment]
@cstyan I don't believe the failing tests are related to my changes...

**Comment by MasslessParticle** [comment]
@chriskuchin they're not related to your changes. Once https://github.com/grafana/loki/pull/8108#issuecomment-1379445593 is merged, you can rebase and this should just work (tm)

**Review by MasslessParticle — APPROVED** [review]

**Comment by MasslessParticle** [comment]
This looks good, thanks for your contribution!

**Pull request merged**